### PR TITLE
[CLOCK] Allow clock to set timezone with autocomplete dropdown option.

### PR DESCRIPTION
**Pull request by dhrubomoy**
Addresses #1273 
- Added `autocomplete` control. To use it just need to pass `options` as either an array of strings or an object with `name` and `value`.
- User will be able to set timezone via autocomplete control. Default timezone is UTC.

## Author Checklist
1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

### CODE REVIEW

File: platform/forms/src/controllers/AutocompleteController.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                if($scope.optionIndex === $scope.filteredOptions.length-1) {
+                    $scope.optionIndex = -1;
+                }
+                $scope.optionIndex++;
+            }
+
+            function fillInputWithString(string) {
+                $scope.hideOptions = true;
+                // Hard coded!!
+                $scope.ngModel[4] = string;
```
Comment (dhrubomoy) [inline platform/forms/src/controllers/AutocompleteController.js]: I was wondering if there was a better way to update the value in text-input (e.g. when clicking on an option)

Comment (larkin) [inline platform/forms/src/controllers/AutocompleteController.js]: @VWoeltjen your thoughts here?

**Comment by larkin** [comment]
@charlesh88 Some CSS changes here that will need your attention.

File: platform/forms/src/controllers/AutocompleteController.js (hunk 1 above)
Comment (VWoeltjen) [inline platform/forms/src/controllers/AutocompleteController.js]: `$scope.ngModel[$scope.field] = string` avoids hard-coding; the property in `ngModel` that the control modifies is passed in as `field`.

Comment (dhrubomoy) [inline platform/forms/src/controllers/AutocompleteController.js]: Aaahhh...how could I have missed that. I used `ngModel[field]` in the template, but completely forgot about it. Thank you so much @VWoeltjen

**Comment by charlesh88** [comment]
Taking a look at a local build of this now.

**Comment by charlesh88** [comment]
@dhrubomoy This is looking really good, and feels great - very sporty! A couple callouts:

- If the user enters text that isn't a valid timezone, and fails to select a zone from the dropdown, then the control silently fails back to UTC. This is concerning from a usability perspective. Could we add validation such that the user must enter a valid timezone before the Ok button is enabled?
- I'd like to address the dropdown arrow:  currently, clicking the arrow doesn't actually do anything. What would be ideal is that clicking the arrow would display the entire list of timezones, regardless of what's currently entered. As soon as the user typed (or deleted chars in the input) then that filtration would take over the list display, as it currently works. This would make the list of zones really discoverable. Alternately, if that's too difficult to do, then we should remove the arrow. 

Again, great work and big thanks. Let me know if there are any questions.

**Comment by dhrubomoy** [comment]
@charlesh88 Thank you so much for your feedback. I will start working on them soon. Changes should be fairly easy to make. The second one should be super easy :)

**Comment by charlesh88** [comment]
@dhrubomoy Reach out to @larkin is you have questions about the validation.

**Comment by dhrubomoy** [comment]
@larkin I have found an ugly way to achieve @charlesh88's point:
> Could we add validation such that the user must enter a valid timezone before the Ok button is enabled?

Setting the `$valid` property of the form disables the OK button. If we add something like following in [`AutocompleteController.js`](https://github.com/dhrubomoy/openmct/blob/3870266131ad792ec292038b72d5cac03d3f8a1d/platform/forms/src/controllers/AutocompleteController.js), it works nicely:
```javascript
if(invalidOptionTyped) {
    $scope.$parent.$parent.$parent.$parent.mctFormInner.$$parentForm.$valid = false;
    // That's why I said its the ugly way to disable the OK button :p
}
```
I was wondering if there is a better way to disable the OK button. For example, how `ng-pattern` works. I just need to pass a regular expression and if the value in input doesn't match the pattern, the OK button will be automatically disabled.

**Comment by dhrubomoy** [comment]
@charlesh88 [commit 3870266](https://github.com/nasa/openmct/pull/1595/commits/3870266131ad792ec292038b72d5cac03d3f8a1d) addresses the second point you made. But the present behavior when clicking the arrow icon is inconsistent with what happens when we click the input. One would display the entire list of timezones and the other would display a list based on what's typed in input. Is this ok or should I make them consistent?

**Comment by charlesh88** [comment]
@dhrubomoy On 
> But the present behavior when clicking the arrow icon is inconsistent with what happens when we click the input. One would display the entire list of timezones and the other would display a list based on what's typed in input. Is this ok or should I make them consistent?

Sorry, but not exactly sure what the question is. This is how I see this working:
- Clicking the arrow should display the complete list regardless of what is entered in the input, always.
- Clicking the input itself (making it the focus) should display a list filtered by what is entered. If nothing is entered, we can display the complete list, or nothing, depending on what's easier. If what is entered doesn't have any matches (like 'asdf') then the nothing displays in the list. This is the current behavior.
- Typing into the input should filter the list on each key down, which is the current behavior. This should take effect regardless of whether the user clicked the arrow initially or the input.
- If a invalid timezone is entered in the input, the form row should be invalid and therefore the form is invalid: the Ok button should be disabled. If a valid timezone is entered, then the form row would be valid. 
- Selecting from the list populates the input with a valid timezone. The user could also manually enter a valid timezone by typing it in, and the form row should be validated in that case as well.

Let me know if there are any questions, and thanks.

**Comment by dhrubomoy** [comment]
@charlesh88 thank you so much for your feedback. Its clear now :) My question was, is it ok if clicking arrow icon and clicking input itself producing two different behavior. But your comment made it clear that it is ok. Thanks again.

Hmm...that's strange! It works in my machine. Here is a screenshot. Did you pull the latest changes from this branch?

autocomplete_icon

**Comment by charlesh88** [comment]
@dhrubomoy That anim looks exactly like what we want. I had pulled latest, but I'll give it a whirl now and see.

**Comment by dhrubomoy** [comment]
Hi @charlesh88, did it work? I just checked checked on another computer and it works as expected!

**Comment by charlesh88** [comment]
@dhrubomoy And seven hours later... sorry! Very cool, dropdown arrow now working great for me.

**Comment by dhrubomoy** [comment]
Hi @larkin, I was wondering if you would be able to help me with disabling the `Ok` button if wrong name is typed. I did some investigation that I mentioned in [this comment](https://github.com/nasa/openmct/pull/1595#issuecomment-306666775), but this is definitely not an ideal solution 😞

**Comment by larkin** [comment]
@dhrubomoy That's a larger issue, and out of scope for this pull request.  For this pull request, don't worry about greying out the `Ok` button.

I've opened https://github.com/nasa/openmct/issues/1623 to track the need to refactor and improve form handling (specification, validation, submission).

**Comment by dhrubomoy** [comment]
@larkin That's great. Thank you so much.
@larkin @charlesh88 @VWoeltjen, Let me know how else I can improve the code and/or if there is any other issues with the change that I should address. Thanks :smiley:

**Comment by charlesh88** [comment]
@dhrubomoy @larkin @VWoeltjen Ok, so where we're at: 
- Dropdown is working the way I'd like it to, awesome.
- We're deferring the form validation per #1623. Not a fan of this approach as we're shipping broken functionality in this interface, but agree that we shouldn't implement a temp solution here and redo later. 

Who's going to do the final merge on this?

**Comment by larkin** [comment]
@VWoeltjen can you do the review here?  You have better knowledge of mct-controls than I do.

File: platform/forms/bundle.js [hunk 2]
```diff
@@ -82,10 +88,21 @@ define([
                         "templateLinker",
                         "controls[]"
                     ]
+                },
+                {
+                    "key": "mctClickElsewhere",
```
Comment (VWoeltjen) [inline platform/forms/bundle.js]: This directive is already defined in [commonUI/general](https://github.com/nasa/openmct/blob/ec4fe8efb39e4decb869edde9c6c61393c32ec6d/platform/commonUI/general/bundle.js#L331-L337); I don't think there is a reason you should need to redefine it here.

File: platform/forms/bundle.js [hunk 3]
```diff
@@ -153,6 +178,13 @@ define([
                         "$scope",
                         "dialogService"
                     ]
+                },
+                {
+                    "key": "mctClickElsewhere",
```
Comment (VWoeltjen) [inline platform/forms/bundle.js]: See above

**Comment by dhrubomoy** [comment]
@VWoeltjen That's great. Thank you so much for your feedback. Just made those changes. Ready for the next ones. Thanks again 😃

**Comment by VWoeltjen** [comment]
Thanks! The `mctClickElsewhere` change looks good, but still seeing errors when I run `gulp checkstyle` or `gulp lint`.

The checkstyle issues can be fixed automatically by `gulp fixstyle`; the lint errors will need to be fixed manually, but these are pretty straightforward:

```
$ gulp lint
[11:41:41] Using gulpfile ~/Projects/WTD/Repositories/openmct/gulpfile.js
[11:41:41] Starting 'lint'...
platform/forms/src/controllers/AutocompleteController.js: line 46, col 19, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 73, col 22, 'fillInputWithIndexedOption' was used before it was defined.
platform/forms/src/controllers/AutocompleteController.js: line 102, col 14, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 112, col 22, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 114, col 14, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 119, col 14, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 124, col 14, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 128, col 14, Missing semicolon.
platform/forms/src/controllers/AutocompleteController.js: line 132, col 14, Missing semicolon.

9 errors
[11:41:53] 'lint' errored after 11 s
[11:41:53] Error in plugin 'gulp-jshint'
Message:
    JSHint failed for: platform/forms/src/controllers/AutocompleteController.js
```

**Comment by dhrubomoy** [comment]
That's good to know. There were lots of missing semicolons :stuck_out_tongue: Just fixed those gulp errors.

**Comment by VWoeltjen** [comment]
Thanks for getting this in quickly! Thought this would be ready for merge, but hitting test failures. I think this is because AutocompleteController constructor takes`$scope, $element` but the spec only provides `$scope`

**Comment by dhrubomoy** [comment]
@VWoeltjen Ahh. My bad. I completely forgot to update and run tests after the code change based on @charlesh88's review. Sorry about that.
Made some changes in [`AutocompleteControllerSpec.js`](https://github.com/nasa/openmct/pull/1595/commits/504b2e1ecf6e55426744e820f7cd2b456e002cbb#diff-df8b85007864fc774a90763aebb5131b) to make the tests pass.

**Comment by dhrubomoy** [comment]
@charlesh88 @VWoeltjen Just had a thought. We can't disable the `Ok` button now. But we can at least show a warning icon if invalid option was typed. After [#1623](https://github.com/nasa/openmct/issues/1623) is fixed, we can remove the icon (only couple of lines of code). My latest change reflect this point. Let me know what you think.

Ran `gulp verify` before pushing the change this time :stuck_out_tongue:

**Comment by VWoeltjen** [comment]
image

This looks a bit inconsistent with the UI to me. Perhaps another option would be to show a small message beneath the control, e.g. "No matching options; defaulting to <option>"? Or we can just leave this for #1623 as discussed above; that would avoid adding something that we'll just pull out later.

@charlesh88 - your thoughts on this?

**Comment by charlesh88** [comment]
@dhrubomoy @VWoeltjen Not a bad idea, but I do agree with Victor on this. I really like the "No matching options" message in the dropdown, but that doesn't help when the user has blurred out of the field; the green checkmark is really the element that is supposed to articulate the validity of the field.

My druthers would be to leave it for now and move on, and pick this up in #1623 as soon as we can.

**Comment by dhrubomoy** [comment]
@VWoeltjen @charlesh88 I completely agree. The UI is inconsistent. Definitely it will leave user confused. Just reverted the commit.

**Comment by VWoeltjen** [comment]
Proceeding with merge now that the UI question is settled. Thank you again for taking the time to work through this issue, it's very much appreciated!

### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

**Pull request merged**